File: russian-roulette/scala/Russian.scala
import java.util.Random
import java.util.Scanner
import java.io.InputStream
import scala.util.control.Breaks._


class Game(var shots: Int) {
  val gun = new Gun(shots)

  val in = new Scanner(System.in)

  def show_prompt() {
    println("Shooting " + gun.current_chamber + " of " + gun.chambers);
    //println(gun.revolver);
    print("Shoot? [y/N] ")
  }

  def start() {
    println("Shooting up to " + gun.chambers);
    //println("Bullet in " + gun.loaded);
    show_prompt();
    breakable {
      while (in.hasNext()) {
        val exp = in.next();
        if (exp == "y" || exp == "Y") {
          println("Shooting..."); 
          gun.shoot;
          if (gun.fired) {
            println("Bang! You're dead!"); break
          } else {
            println("Whew! No bullet.")
          }
        } else {
          println("Chicken! Ending game."); break
        }
        show_prompt()
      }
    }
  }
}

class Gun(var shots: Int) {
  val loaded = Math.abs(new Random().nextInt() % shots)

  val revolver = (0 until shots).map(x => x == loaded)

  var current = 0

  def chambers = revolver.size

  def fired : Boolean = revolver(current - 1)

  def shoot {
    current = current + 1
  }

  def current_chamber = current + 1
}

object Main extends App {
  val game = new Game(6)
  game.start()
}

File: russian-roulette/python/Russian.py
import random

class Gun:
    def __init__(self, shots):
        self.shots = shots
        self.current = 0
        self.chamber = random.randint(1, shots)
        #print "Loaded chamber", self.chamber

    def loaded_chamber(self):
        return self.chamber

    def shoot(self):
        self.current += 1
        return self.current == self.chamber


def prompt():
    response = raw_input("Shoot? [Y/n] ")
    return response == 'n' or response == 'N'


def start():
    shots = int(input("How many shots? "))
    if shots < 1:
        print "Shots must be greater than zero"
        return start()
    return shots


def play():
    shots = start()
    print "Shooting", shots, "shots."
    gun = Gun(shots)
    for x in range(shots+1):
        print "Shot number", x+1
        stop = prompt()
        if stop:
            print "Chicken!", "The bullet was in", gun.loaded_chamber()
            break
        if gun.shoot():
            print "Bang! You're dead!"
            break
    else:
        print "How are you not dead?"
    

def main():
    print "Welcome to Russian Roulette"
    play()

if __name__ == '__main__':
    main()